### ISSUE TITLE
Error rendering docstrings

### ISSUE THREAD

**Issue by yiquintero** [tudat-team/tudatpy#202]
## Description

The docstrings for the functions listed below are incorrectly rendered by Sphinx. The issue occurs when building the documentation locally using tudat-multidoc or the new method introduced as part of the [2024 updates to the tudatpy API documentation workflow](https://github.com/tudat-team/tudatpy/wiki/2024-Updates-to-tudatpy-API-Documentation-Workflow). The docstrings are also rendered incorrectly on the [API Reference website](https://py.api.tudat.space/en/latest/).

### Ephemeris
1. interpolated_spice
2. tabulated_from_existing

### Estimation
<details>
<summary> List of Functions </summary>

1. add_dependent_variable
2. compatible_dependent_variable_settings
3. compatible_dependent_variables_list
4. concatenated_dependent_variable
5. dependent_variable
6. dependent_variable_history
7. dependent_variable_history_per_set
8. filter_observations
9. get_bodies_in_link_ends
10. get_computed_observations
11. get_concatenated_computed_observations
12. get_concatenated_float_observation_times
13. get_concatenated_observation_times
14. get_concatenated_observations
15. get_concatenated_observations_and_times
16. get_concatenated_residuals
17. get_concatenated_weights
18. get_mean_residuals
19. get_observable_types
20. get_observation_times
21. get_observations
22. get_observations_and_times
23. get_reference_points_in_link_ends
24. get_residuals
25. get_rms_residuals
26. get_single_observation_sets
27. get_time_bounds_list
28. get_time_bounds_per_set
29. get_weights
30. set_constant_weight (x2)
31. set_reference_point (x2)
32. set_reference_points
33. set_tabulated_weights
34. split_observation_sets
35. EstimationInput constructor

</details>

### Observation
<details>
<summary> List of Functions </summary>

1. two_way_doppler_averaged
2. one_way_range
3. n_way_range
4. two_way_range
5. angular position
6. relative_angular_position
7. one_way_doppler_instantaneous
8. one way doppler averaged
9. n_way_doppler_averaged
10. n_way_doppler_averaged_from_one_way_links

</details>

### Propagator
<details>
<summary> List of Functions </summary>

1. hybrid_arc
2. translational
3. rotational
4. mass
5. multitype
6. multi_arc

</details>

**Comment by larshinueber** [comment]
Hi all,

Based on the [pybind docs](https://pybind11.readthedocs.io/en/stable/advanced/functions.html#default-arguments-revisited) this can be resolved using `py::arg_v` and manually specifying the default argument as string.
For the ephemeris, this would look as follows:
![Image](https://github.com/user-attachments/assets/8631175b-8ae9-483a-a8dc-f82436d71b24)

However, this breaks a bit with the current way of instantiating interpolator settings (or settings in general), since they are typically not created through the init method of the class itself, but through a factory function.
Translating the instantiation from the init method to the factory function notation does not seem like a better option, since it would mean that the developer/person documenting the function has to reverse engineer the factory function and use it in the docstring, which seems very error-prone.

What are your thoughts?

**Comment by DominicDirkx** [comment]
Thanks for figuring out a potential way to solve this!

As I understand it, we then have three options, keeping the following in mind:

> Translating the instantiation from the init method to the factory function notation does not seem like a better option, since it would mean that the developer/person documenting the function has to reverse engineer the factory function and use it in the docstring, which seems very error-prone

1. Have the C++-y default arguments, which can look horrible and don't really add any information to the user
2. Use the init function of the objects, which typically us not used/exposed
3. Use the factory function, which would need to be correctly hard-coded as a string by the developer. 

Right?

**Comment by larshinueber** [comment]
Yes exactly. There might be two more options: defining the default argument from the factory function instead of the class constructor (4) or defining `__repr__` methods (5).

Maybe it's easiest to demonstrate with code examples:

<details><summary>Option 1: C++-defaulty arguments (status-quo)</summary>
<p>

Exposure code:
```c++
m.def( "tabulated_from_existing",
           /*
           remaining exposure
           */
           py::arg( "interpolator_settings" ) = std::make_shared< ti::LagrangeInterpolatorSettings >( 8 ),
```
will be displayed as:

![Image](https://github.com/user-attachments/assets/4312ca2f-bad4-4582-8f45-70bebf3cd841)

Advantage:
- No extra effort in implementation

Disadvantage:
- No information for users, parameters need to be explicitly stated in docstrings

</p>
</details> 

<details><summary>Option 2: Default argument from class constructor, manual string representation by developer</summary>
<p>

```c++
m.def( "tabulated_from_existing",
           /*
           remaining exposure
           */
           py::arg_v( "interpolator_settings", std::make_shared< ti::LagrangeInterpolatorSettings >( 8 ), "LagrangeInterpolatorSettings(8)" ),

```
will be displayed as:

![Image](https://github.com/user-attachments/assets/f58e5d76-d38f-48c7-96b9-a8e1a47d93ea)

Advantage:
- Little extra effort to implement

Disadvantage:
- The init methods are not always exposed, since the creation of objects through factory functions is preferred (in this case lagrange_interpolation())
- Since only the 8th order is specified in the default argument (`LagrangeInterpolatorSettings`), the object will be created with the remaining C++ default arguments (`selectedLookupScheme`, `lagrangeBoundaryHandling`, etc.). Even if the init method of LagrangeInterpolatorSettings was exposed in Python, the unspecified default arguments are not necessarily the same as on the C++ side.

</p>
</details> 

<details><summary>Option 3: Default argument from class constructor, manual string representation by developer but translated to exposed python factory function</summary>
<p>

```c++
m.def( "tabulated_from_existing",
           /*
           remaining exposure
           */
           py::arg_v( "interpolator_settings", std::make_shared< ti::LagrangeInterpolatorSettings >( 8 ), "lagrange_interpolation(8)" ),

```
I realized this doesn't make sense, if the string representation should be the Python exposure factory function option 4 should be used.

</p>
</details> 

<details><summary>Option 4: Default argument from factory function, manual string representation by developer</summary>
<p>

```c++
m.def( "tabulated_from_existing",
           /*
           remaining exposure
           */
           py::arg_v( "interpolator_settings", ti::lagrangeInterpolation( 8 ), "lagrange_interpolation(8)" ),

```
will be displayed as:

![Image](https://github.com/user-attachments/assets/c4788ddf-3a1e-4180-a813-7fc53ca96042)

Advantage:
- Modest extra effort to implement
- The default argument representation is the function that the user will be able to use, since the factory functions are exposed in most cases

Disadvantage:
- Similar to option 2: Since only the 8th order is specified in the default argument (`lagrangeInterpolation`), the object will be created with the remaining C++ default arguments (`selectedLookupScheme`, `lagrangeBoundaryHandling`, etc.). The unspecified default arguments of the Python exposed function are not necessarily the same as on the C++ side. See for an example `estimation_convergence_checker`, which is exposed with `py::arg( "minimum_residual" ) = 0.0` (see [expose_estimation_propagated_covariance.cpp#L564](https://github.com/tudat-team/tudatpy/blob/ff783c32de9b5438f3390c82585bf92658f29657/tudatpy/kernel/expose_numerical_simulation/expose_estimation_propagated_covariance.cpp#L564)), but has `const double minimumResidual = 1.0E-20,` specified on the C++ side ([podInputOutputTypes.h#L670](https://github.com/tudat-team/tudat/blob/3c7d26f864fd04559a2c5cc26fcf431bbbced194/include/tudat/astro/orbit_determination/podInputOutputTypes.h#L670)). In this case, the developer would have to manually compare the C++ default constructor argument with the Python exposure, spot the difference and add it to the string representation to specify the default arguments correctly.

</p>
</details> 

<details><summary>Option 5: Define `__repr__` functions for default argument classes</summary>
<p>

Since the object representations are taken from the `__repr__` methods, these could be redefined as per [the official docs](https://pybind11.readthedocs.io/en/stable/advanced/functions.html#default-arguments-revisited). This would be the cleanest way, as the string representation can access the internal attribute, but would require the most implementation effort.

</p>
</details> 

In summary, I think the difficulty comes from the fact that the default arguments in the C++ constructors and the Python exposures are not necessarily the same.

For the time being, I think option 4 would have the best effort:impact ratio, since it will be displayed in the same way as users would use in their own code. In the long term option 5 could be interesting, as it would improve the output when calling print() on an object, however, this comes at a significant implementation effort.

(Wow, this turned longer than I expected for a relatively small problem. 😅)
